proof/crefine/AARCH64/Fastpath_C.thy

```diff
@@ -1736,776 +1684,989 @@ lemma recv_ep_queued_st_tcb_at':
+                              apply vcg
+
+                             apply (rule conseqPre, vcg, clarsimp)
+                            apply (simp add: cte_level_bits_def field_simps shiftl_t2n
+                                             ctes_of_Some_cte_wp_at
+                                        del: all_imp_to_ex cong: imp_cong conj_cong)
+                            apply (wp hoare_vcg_all_lift threadSet_ctes_of
+                                      hoare_vcg_imp_lift' threadSet_valid_objs'
+                                      threadSet_st_tcb_at_state threadSet_cte_wp_at'
+                                      threadSet_cur asid_has_vmid_lift
+                                    | simp add: cur_tcb'_def[symmetric]
+                                    | strengthen not_obj_at'_strengthen)+
+                           apply (vcg exspec=thread_state_ptr_set_tsType_np_modifies)
+                          apply (wp hoare_vcg_all_lift threadSet_ctes_of
+                                    hoare_vcg_imp_lift' threadSet_valid_objs'
+                                    threadSet_st_tcb_at_state threadSet_cte_wp_at'
+                                    threadSet_cur
+                                  | simp add: cur_tcb'_def[symmetric])+
+                            apply (simp add: valid_tcb'_def tcb_cte_cases_def
+                                             valid_tcb_state'_def cteSizeBits_def)
+                            apply (wp hoare_vcg_all_lift hoare_vcg_imp_lift
+                                      set_ep_valid_objs' asid_has_vmid_lift
+                                      setObject_no_0_obj'[where 'a=endpoint, folded setEndpoint_def]
+                                   | strengthen not_obj_at'_strengthen)+
+                             apply (rule_tac Q="\<lambda>_ s. hd (epQueue send_ep) \<noteq> curThread
+                                                      \<and> pred_tcb_at' itcbState ((=) (tcbState xa)) (hd (epQueue send_ep)) s"
+                                          in hoare_post_imp)
+                              apply fastforce
+                             apply wp+
+                         apply simp
+                         apply (vcg exspec=endpoint_ptr_mset_epQueue_tail_state_modifies
+                                  exspec=endpoint_ptr_set_epQueue_head_np_modifies)
+                        apply simp
+                        apply (vcg exspec=cap_endpoint_cap_get_capCanGrant_modifies
+                                   exspec=cap_endpoint_cap_get_capCanGrantReply_modifies)
+                       apply simp
+                       (* throw away results of isHighestPrio and the fastfail shortcut *)
+                       apply (wp (once) hoare_drop_imp, wp)
+                       apply (wp (once) hoare_drop_imp, wp)
+
+(* FIXME AARCH64 this is frustrating:
+   - can't use clarsimp because it'll introduce new free variables (\<exists>_. _ = Some _) that will
+     cause schematic unification problems
```

Comment:
We could try to remove the simp rule that turns None into Ex Some. With that, clarsimp might be fine.

Reply:
There's still the issue of pairs, which would need to be folded back into `fst` and `snd`. Do we have tech to handle that?
There's also what Corey wrote on the wp schematic issue, where we might have a wrong rule in the `wp_comb` set.

Reply:
To clarify, if I'm right in what I thought was going on with the `wp_comb` rule then it would just help with the problem you ran into of a schematic assumption being produced and then unified with `False`. The actual proof would still require either your careful rule application or the careful adjustment of the simp set like you two are talking about.

Reply:
Pairs being split should be handled correctly by `wpfix`, so I think removing the None/Some rule has a chance in this case (it won't fix the general problem). It has a tag already, so we're fine to merge, I was just curious what people's thoughts were on this one. If I get to it in the cleanup run I might try out removing the rule and see what happens.

Reply:
Ok, that would be nice, but I think the wp_comb thing also should happen if we're doing cleanup at some point, because the combo of wp, simp, vcg is a disaster to work with otherwise. When you say "when I get to it in the cleanup run", does that mean you've assumed responsibility for the doing entire cleanup run?

Reply:
I was planning to volunteer to do a first pass, but I wouldn't say I have assumed responsibility for all of the cleanup :-) There will probably be a whole bunch of things that need some insight and where we might want to discuss more, it'd be nice to do these together in some way.